crates/telio-starcast/src/transport.rs

```diff
@@ -0,0 +1,231 @@
+    }
+
+    /// Stop the transport component
+    pub async fn stop(self) {
+        let _ = self.task.stop().await.resume_unwind();
+    }
+}
+
+struct State {
+    transport_socket: Arc<UdpSocket>,
+    packet_chan: Chan<Vec<u8>>,
+    recv_buffer: Vec<u8>,
+    nat: StarcastNat,
+    peers: Vec<Peer>,
+    multicast_ips: Vec<IpNetwork>,
+}
+
+impl State {
+    fn configure(&mut self, config: Config) {
+        self.peers = config.peers;
+    }
+
+    async fn handle_local_multicast_packet(&mut self, packet: Vec<u8>) -> Result<(), Error> {
+        let failed_peers = join_all(self.peers.iter().map(|peer| {
+            self.transport_socket
+                .send_to(
+                    &packet,
+                    SocketAddr::new(peer.addr, MULTICAST_TRANSPORT_PORT),
+                )
+                .map_err(|_| peer.public_key)
+        }))
+        .await
+        .into_iter()
+        .filter_map(|res| match res {
+            Ok(_) => None,
+            Err(pk) => Some(pk),
+        })
+        .collect::<Vec<_>>();
+
+        if failed_peers.is_empty() {
+            Ok(())
+        } else {
+            Err(Error::FanoutFailed(failed_peers))
+        }
+    }
+
+    async fn handle_mapped_unicast_packet(&mut self, mut packet: Vec<u8>) -> Result<(), Error> {
+        self.nat
+            .translate_outgoing(&mut packet)
+            .map_err(Error::NatError)?;
+        self.transport_socket
+            .send(&packet)
+            .await
+            .map(|_| ())
+            .map_err(|_| Error::SocketSendError)
+    }
+
+    async fn handle_remote_packet(&mut self, mut packet: Vec<u8>) -> Result<(), Error> {
+        self.nat
+            .translate_incoming(&mut packet)
+            .map_err(Error::NatError)?;
+        self.packet_chan
+            .tx
+            .send(packet)
+            .await
+            .map_err(|_| Error::ChannelSendError)
+    }
+
+    fn has_multicast_dst(&self, packet: &mut [u8]) -> Result<bool, Error> {
+        let dst = if let Some(packet) = MutableIpv4Packet::new(packet) {
+            IpAddr::V4(packet.get_destination())
+        } else if let Some(packet) = MutableIpv6Packet::new(packet) {
+            IpAddr::V6(packet.get_destination())
+        } else {
+            return Err(Error::InvalidIpPacket);
+        };
+        Ok(self
+            .multicast_ips
+            .iter()
+            .any(|network| network.contains(dst)))
+    }
+}
+
+#[async_trait]
+impl Runtime for State {
+    const NAME: &'static str = "StarcastTransport";
+
+    type Err = Error;
+
+    async fn wait(&mut self) -> WaitResponse<'_, Self::Err> {
+        let res = tokio::select! {
+            Some(mut packet) = self.packet_chan.rx.recv() => {
+                match self.has_multicast_dst(&mut packet) {
+                    Ok(val) if val => self.handle_local_multicast_packet(packet).await,
+                    Ok(_) => self.handle_mapped_unicast_packet(packet).await,
+                    Err(e) => Err(e),
```

Comment:
Do we want to stop entire transport task if error is encountered here?
I think we should just log it.